File: lib/gari/layer.c
#include "gari.h"

#define GARI_INTERN_ONLY
#include "gari_intern.h"

/* Layers in a tile map. */

struct GariLayer_ {
  GariCamera  *   camera;
  GariTileset *   set;
  GariTile    *** tiles;
  GariImage   *   background;
  int         **  tilesave;
  int             tilewide; // width of map tiles in pixels
  int             tilehigh; // height of map tiles in pixels
  int             gridwide; // width of the tile map in grid points (tiles)
  int             gridhigh; // height of the tile map in grid points (tiles)
  int             realwide; // width of whole tile map in pixels
  int             realhigh; // height of whole tile map in pixels  
};

// cleans up a layer by freeing the memory references it holds internally,
// but does not free the layer itself 
GariLayer * gari_layer_done(GariLayer * layer) {
  int index;
  // First free the rows.
  if (!layer) return NULL;
  for (index = 0; index < layer->gridhigh ; index ++) {
    GARI_FREE(layer->tiles[index]);
    GARI_FREE(layer->tilesave[index]);
  }
  // Then the column holder arrays. 
  GARI_FREE(layer->tiles);
  GARI_FREE(layer->tilesave);
  // Size is now zero.
  layer->gridhigh = 0;
  layer->gridwide = 0;
  return layer;  
}

GariLayer * gari_layer_init(GariLayer * layer, GariTileset * set, 
                           int gridwide, int gridhigh, 
                           int tilewide, int tilehigh) {
  int index;
  if((!layer) || (!set)) return NULL;
  layer->set        = set;
  layer->gridwide   = gridwide;
  layer->gridhigh   = gridhigh;
  layer->tilewide   = tilewide;  
  layer->tilehigh   = tilehigh;
  layer->realwide   = layer->tilewide * layer->gridwide;
  layer->realhigh   = layer->tilehigh * layer->gridhigh;
  // declare background, etc, empty
  layer->camera     = NULL;
  layer->background = NULL;
  
  // Precalculate dimensions...  
  // And allocate space for the tiles and tile indices.
  layer->tiles    = GARI_MALLOC(sizeof(GariTile**) * layer->gridhigh);
  if(!layer->tiles) {
    return NULL;
  }   
  // First empty...
  for (index = 0; index < layer->gridhigh ; index ++) {
    layer->tiles[index] = NULL; 
  }
  
  // Then allocate column rows, calling gari_layer done on failiure
  // that's why nwe neede dto null everything first) 
  for (index = 0; index < layer->gridhigh ; index ++) {
    layer->tiles[index] = GARI_MALLOC(sizeof(GariTile *) * layer->gridwide);
    if(!layer->tiles[index]) {
      gari_layer_done(layer);
      return NULL;
    } 
  }
  // Now also the editing support matrix...
  layer->tilesave = GARI_MALLOC(sizeof(int*) * layer->gridhigh);
  if(!layer->tilesave) { // free up again on failiure
    gari_layer_done(layer);
    return NULL;
  }
  // And the rows of the editing matrix.
  for (index = 0; index < layer->gridhigh ; index ++) {
    layer->tilesave[index] = GARI_MALLOC(sizeof(int) * layer->gridwide);
    if(!layer->tilesave[index]) {
      gari_layer_done(layer);
      return NULL;
    }
  }
  
  return layer;
}

/** Returns TRUE if the given gridx and gridy are outside the grid
 Returns FALSE if inside the grid.
*/
int gari_layer_outsidegrid(GariLayer * layer, int gridx, int gridy) {
  if (!layer) return TRUE;
  if ((gridx < 0) || (gridy < 0)) return TRUE;
  if ((gridx >= layer->gridwide) || (gridx >= layer->gridhigh)) return TRUE;
  return FALSE;
}

/** Sets the tile in this layer to the tile with this tile index in the 
layer's Tileset. Returns the tile thus set or NULL on error, or if a NULL 
tile was set (index -1).  */
GariTile * gari_layer_set(GariLayer * layer, int gridx, int gridy, int tileid) { 
  GariTile * realtile = NULL;
  if (gari_layer_outsidegrid(layer, gridx, gridy)) return NULL;                 
  if (tileid < 0) {
    realtile = NULL;
    tileid   = -1; 
  } else {
    realtile = gari_tileset_get(layer->set, tileid);
    if(!realtile) return NULL;
    // bail out if no such tile found. 
  }
  layer->tiles[gridy][gridx]    = realtile;
  layer->tilesave[gridy][gridx] = tileid;
  return realtile;
}  

/** Returns the tile in the layer's grid at the given grid coordinates,
* returns -2 if the fcoordinates are out of bounds, -1 if it was an empty tile.  
*/
int gari_layer_get(GariLayer * layer, int gridx, int gridy) { 
  GariTile * realtile = NULL;
  if (gari_layer_outsidegrid(layer, gridx, gridy)) return -2;
  return layer->tilesave[gridy][gridx];
}  


/*
  
  # Returns the actual tile at the given x and y coordinates, 
  # or nil if the coordinates are out of bounds
  def get_tile(x, y) 
    gx = x; gy = y
    return nil if gy > @high or gx > @wide or gy < 0 or gx < 0  
    return @layer[gy][gx]
  end
  
  # Iterate over all tiles in this layer, column by column. 
  # Yields Tile objects, and the world x and y coordinate of this tile 
  # in the layer.
  def each_tile()
    for y in (0...@high)
      for x in (0...@wide)          
          wx, wy = tile_to_world(x, y)
          yield @layer[y][x], wx, wy
      end
    end
  end

  def draw(screen, x, y)
    txstart     = ( x / @tilewide )
    tystart     = ( y / @tilehigh )
    xtilestop   = (screen.w / @tilewide) + 1
    ytilestop   = (screen.h / @tilehigh) + 1
    txstop      = xtilestop + txstart
    tystop      = ytilestop + tystart
    drawx       = 0
    drawy       = 0
    row         = nil
    aidimage    = nil
    nilcount    = 0
    return if (txstart >= @wide or tystart >= @high) 
    txstart = 0    if (txstart < 0) 
    tystart = 0    if (tystart < 0)
    txstop  = @wide if (txstop > @wide) 
    tystop  = @high if (tystop > @high) 
    drawy   = -y + ( (tystart-1) * @tilehigh )
    tydex   = tystart
    # for tydex in (tystart...tystop)
    while tydex < tystop do 
      drawy   += @tilehigh;
      drawx   = -x + ( (txstart-1) * @tilewide )
      row     = @layer[tydex]
      # for txdex in (txstart...txstop) do
      txdex = txstart
      while txdex < txstop
        drawx   += @tilewide        
        aidtile  = row[txdex]
        unless aidtile.nil? then
          aidimage    = aidtile.get
          SDL::Surface.blit(aidimage, 0,0,0,0, screen, drawx, drawy)
        end
        txdex += 1
      end
      tydex  += 1
    end
  end
*/

File: lib/rogaru/tilemap/layer.rb


class Rogaru::Tilemap::Layer
  attr_reader   :tilewide
  attr_reader   :tilehigh
  attr_reader   :layer
  attr_accessor :tileset
    
  def initialize(tileset, wide, high) 
    @tileset    = tileset 
    @layer      = []
    @layer_save = []
    @high       = high
    @wide       = wide
    for index in (0...@high)
      @layer[index] = []
      @layer_save[index] = []
      for jdex in (0...@wide)
        @layer[index].push(nil)
        @layer_save[index].push(-1)
      end  
    end
    @tilehigh   = @tileset.tilehigh
    @tilewide   = @tileset.tilewide
    @realhigh   = @tilehigh * @high
    @realwide   = @tilewide * @wide
    @x          = 0
    @y          = 0
    @vx         = 0
    @vy         = 0
  end
  
  def set_tiles(tilearray) 
    @tiles = tilearray
  end

  def add_tile(tile) 
    raise "Obsolete"
  end

  # Sets the tile in this layer to the tile with this tile index
  def set(x, y, tile_id)
    if (x < 0) or (y<0) or ( x >= @wide) or (y >= @high)
      raise "Index #{x} #{y} #{tile_id} not valid." 
    end  
    real_tile_id    = -1
    real_tile       = nil
    unless tile_id.nil? or tile_id < 0 
      real_tile       = @tileset.get_tile(tile_id)
      real_tile_id    = tile_id
      unless real_tile
        warn "No real tile found for #{x} #{y} #{tile_id}" 
        real_tile_id    = -1
        real_tile       = nil
      end
    end
    @layer[y][x]      = real_tile 
    @layer_save[y][x] = real_tile_id
  end
  
  # Gets the tile index of the tile at the given x,y coordinates,
  # or -1 if out of bounds 
  def get(x, y) 
    gx = x; gy = y
    return -1 if gy > @high or gx > @wide or gy < 0 or gx < 0  
    return @layer_save[gy][gx]
  end
  
  # Returns the actual tile at the given x and y coordinates, 
  # or nil if the coordinates are out of bounds
  def get_tile(x, y) 
    gx = x; gy = y
    return nil if gy > @high or gx > @wide or gy < 0 or gx < 0  
    return @layer[gy][gx]
  end
  
  # Iterate over all tiles in this layer, column by column. 
  # Yields Tile objects, and the world x and y coordinate of this tile 
  # in the layer.
  def each_tile()
    for y in (0...@high)
      for x in (0...@wide)          
          wx, wy = tile_to_world(x, y)
          yield @layer[y][x], wx, wy
      end
    end
  end

  def draw(screen, x, y)
    txstart     = ( x / @tilewide )
    tystart     = ( y / @tilehigh )
    xtilestop   = (screen.w / @tilewide) + 1
    ytilestop   = (screen.h / @tilehigh) + 1
    txstop      = xtilestop + txstart
    tystop      = ytilestop + tystart
    drawx       = 0
    drawy       = 0
    row         = nil
    aidimage    = nil
    nilcount    = 0
    return if (txstart >= @wide or tystart >= @high) 
    txstart = 0    if (txstart < 0) 
    tystart = 0    if (tystart < 0)
    txstop  = @wide if (txstop > @wide) 
    tystop  = @high if (tystop > @high) 
    drawy   = -y + ( (tystart-1) * @tilehigh )
    tydex   = tystart
    # for tydex in (tystart...tystop)
    while tydex < tystop do 
      drawy   += @tilehigh;
      drawx   = -x + ( (txstart-1) * @tilewide )
      row     = @layer[tydex]
      # for txdex in (txstart...txstop) do
      txdex = txstart
      while txdex < txstop
        drawx   += @tilewide        
        aidtile  = row[txdex]
        unless aidtile.nil? then
          aidimage    = aidtile.get
          screen.blit(drawx, drawy, aidimage)
        end
        txdex += 1
      end
      tydex  += 1
    end
  end

  def draw_effects(screen, x, y)
    txstart     = ( x / @tilewide )
    tystart     = ( y / @tilehigh )
    xtilestop   = (screen.w / @tilewide) + 1
    ytilestop   = (screen.h / @tilehigh) + 1
    txstop      = xtilestop + txstart
    tystop      = ytilestop + tystart
    drawx       = 0
    drawy       = 0
    row         = nil
    aidimage    = nil
    nilcount    = 0
    return if (txstart >= @wide or tystart >= @high) 
    txstart = 0    if (txstart < 0) 
    tystart = 0    if (tystart < 0)
    txstop  = @wide if (txstop > @wide) 
    tystop  = @high if (tystop > @high) 
    drawy   = -y + ( (tystart-1) * @tilehigh )
    c1      = screen.format.map_rgb(10,10,10)
    c2      = screen.format.map_rgb(255,0,0)
    c3      = screen.format.map_rgb(0,255,255)
    c4      = screen.format.map_rgb(255,255,0)
    c5      = screen.format.map_rgb(255,255,255)
    c6      = screen.format.map_rgb(255,255,255)
    
    for tydex in (tystart...tystop) do 
      drawy   += @tilehigh;
      drawx   = -x + ( (txstart-1) * @tilewide )
      row     = @layer[tydex]
      for txdex in (txstart...txstop) do
        drawx   += @tilewide        
        aidtile  = row[txdex]
        unless aidtile.nil? then
          screen.draw_rect(drawx   , drawy    , 2, 2, c1, true) unless aidtile.walk
          screen.draw_rect(drawx   , drawy + 2, 2, 2, c2, true) if aidtile.swim
          screen.draw_rect(drawx   , drawy + 4, 2, 2, c3, true) if aidtile.pain
          screen.draw_rect(drawx+8 , drawy    , 2, 2, c4, true) unless aidtile.jump
          screen.draw_rect(drawx+12, drawy    , 2, 2, c5, true) if aidtile.ledge
          screen.draw_rect(drawx+16, drawy    , 2, 2, c6) if aidtile.up
          screen.draw_rect(drawx+16, drawy+16 , 2, 2, c6) if aidtile.down
        end
      end
    end
  end

  def w
    return @realwide
  end

  def h
    return @realhigh
  end
  
  def tw
    return @wide
  end

  def th
    return @high
  end

  def to_xml() 
    layer = REXML::Element.new 'layer'
    sizehash = {'wide' => @wide.to_s, 
          'high' => @high.to_s,
          'tilewide' => @tilewide.to_s, 
          'tilehigh' => @tilehigh.to_s }
          
    layer.add_attributes(sizehash)
    rows = layer.add_element 'rows'
    for y in (0...@high) do
      row = rows.add_element('row')
      row.add_attribute('y', y.to_s) 
      row.text = @layer_save[y].join(' ');      
    end
    return layer
  end
  
  def self.new_from_xml(xml_layer, tileset)
    mw    = xml_layer.attributes['wide'].to_i
    mh    = xml_layer.attributes['high'].to_i
    tw    = xml_layer.attributes['tilewide'].to_i
    th     = xml_layer.attributes['tilehigh'].to_i
    layer = Tilemap::Layer.new(tileset, mw, mh)
    xml_layer.elements.each('rows/row') do | xml_row |
      y   = xml_row.attributes['y'].to_i
      row = xml_row.text.tr("\n\t\r,;", "     ").strip.split(" ").map { |v| v.to_i  }
      # parse the row, translate all seps to spaces, strip the edges, and split on space
      # and finally transfor into an array of integers.
      stop = row.size
      for x in (0...stop) do
        layer.set(x, y, row[x])
      end
    end
    return layer
  end
    
  # Loads a layer from a tmx file (www.mapeditor.org)
  def self.new_from_tmx(tmx, tileset)
    mw    = tmx.attributes['width'].to_i
    mh    = tmx.attributes['height'].to_i
    layer = Tilemap::Layer.new(tileset, mw, mh)
    y     = 0
    x     = 0
    tmx.elements.each('data/tile') do | tmx_tile |
      unless x < mw
        x   = 0
        y  += 1
      end
      index = tmx_tile.attributes['gid'].to_i 
      index = -1 if index < 1 
      layer.set(x, y, index)
      x    += 1
    end
    return layer
  end

  def screen_to_tile_x(top_x, screen_x)
    return ((top_x + screen_x) / @tilewide).floor
  end 
  
  def screen_to_tile_y(top_y, screen_y)
    return ((top_y + screen_y) / @tilehigh).floor
  end 

  def tile_to_screen_x(top_x, tile_x)
    return   (tile_x * @tilewide) - top_x
  end 
  
  def tile_to_screen_y(top_y, tile_y)
    return   (tile_y * @tilehigh) - top_y
  end 

  def tile_to_world(tile_x, tile_y)
    return tile_x * @tilewide, tile_y * @tilehigh
  end

end

